### PR TITLE
docs: add exceptions API section

### PR DESCRIPTION
**Pull request by aarnphm**
Add a quick reference to our exceptions.

Signed-off-by: Aaron <29749331+aarnphm@users.noreply.github.com>

### CODE REVIEW

File: src/bentoml/exceptions.py [hunk 1]
```diff
@@ -18,7 +17,7 @@ def __init__(self, message: str):
 
 class StateException(Exception):
     """
-    Raise when the state of an object is not valid
+    Raise when the state of an object is not valid.
```
Comment (sauyon) [inline src/bentoml/exceptions.py:20]: Maybe we should reword all of these from "Raise" to "Raised"?

File: src/bentoml/exceptions.py [hunk 2]
```diff
@@ -95,10 +95,13 @@ class BentoMLConfigException(BentoMLException):
 
 class MissingDependencyException(BentoMLException):
     """
-    Raise when BentoML component failed to load required dependency - some BentoML
-    components has dependency that is optional to the library itself. For example,
-    when using SklearnModel, the scikit-learn module is required although
-    BentoML does not require scikit-learn to be a dependency when installed
+    Raise when BentoML component failed to load required dependency.
+
+    Some BentoML components has optional dependencies that can be installed as extensions.
```
Comment (sauyon) [inline src/bentoml/exceptions.py:100]:
```suggestion
    Some BentoML components have optional dependencies that can be installed as extensions.
```

File: src/bentoml/exceptions.py [hunk 3]
```diff
@@ -95,10 +95,13 @@ class BentoMLConfigException(BentoMLException):
...
+
+    For example, when using the :class:`~bentoml._internal.io_descriptors.json.JSON` IODescriptor,
+    ``pydantic`` is considered as an optional feature if users want to use it to validate. BentoML
+    will still works without ``pydantic`` installed.
```
Comment (sauyon) [inline src/bentoml/exceptions.py:104]:
```suggestion
    will still work without ``pydantic`` installed.
```

File: src/bentoml/exceptions.py [hunk 4]
```diff
@@ -107,8 +110,12 @@ class CLIException(BentoMLException):
 
 
 class YataiRESTApiClientError(BentoMLException):
+    """Raise when there is an error during push/pull bentos/models."""
+
     pass
 
 
 class ImportServiceError(BentoMLException):
```
Comment (sauyon) [inline src/bentoml/exceptions.py:118]: "failed to import the user's service file" probably is more accurate than "service.py".

File: src/bentoml/exceptions.py [hunk 5]
```diff
@@ -107,8 +110,12 @@ class CLIException(BentoMLException):
 
 
 class YataiRESTApiClientError(BentoMLException):
+    """Raise when there is an error during push/pull bentos/models."""
```
Comment (sauyon) [inline src/bentoml/exceptions.py:113]: Maybe something like "when communicating with the Bento Cloud server"?

**Review by sauyon — CHANGES_REQUESTED** [review]

File: src/bentoml/exceptions.py [hunk 6]
```diff
@@ -36,79 +35,87 @@ def __init__(self, message: str, payload: BentoMLException | None = None):
...
 
 
 class YataiRESTApiClientError(BentoMLException):
+    """Raised when communicating with BentoCloud server."""
+
     pass
 
 
 class ImportServiceError(BentoMLException):
+    """Raised when failed to import the user's service file."""
```
Comment (sauyon) [inline src/bentoml/exceptions.py:119]:
```suggestion
    """Raised when BentoML failed to import the user's service file."""
```

**Review by sauyon — APPROVED** [review]
Last thing I think.

**Comment by aarnphm** [comment]
@sauyon PTAL.

**Review by sauyon — APPROVED** [review]

**Pull request merged**